docs/nnx/surgery.md

```diff
@@ -0,0 +1,261 @@
+
+# Make your local change on the checkpoint
+raw = checkpointer.restore('/tmp/nnx-surgery-state')
+pprint(raw)
+raw['layer1'], raw['layer2'] = raw['linear1'], raw['linear2']
+del raw['linear1'], raw['linear2']
+
+restored_model = module_from_variables_dict(
+  lambda: nnx.eval_shape(lambda: ModifiedTwoLayerMLP(4, rngs=nnx.Rngs(0))),
+  raw,
+  lambda path: path[:-1] if path[-1] == 'raw_value' else path
+)
+
+np.testing.assert_allclose(restored_model(jnp.ones((3, 4))), old_model(jnp.ones((3, 4))))
+```
+
+    {'linear1': {'bias': {'raw_value': Array([0., 0., 0., 0.], dtype=float32)},
+                 'kernel': {'raw_value': Array([[-0.80345297, -0.34071913, -0.9408296 ,  0.01005968],
+           [ 0.26146442,  1.1247735 ,  0.54563737, -0.374164  ],
+           [ 1.0281805 , -0.6798804 , -0.1488401 ,  0.05694951],
+           [-0.44308168, -0.60587114,  0.434087  , -0.40541083]],      dtype=float32)}},
+     'linear2': {'bias': {'raw_value': Array([0., 0., 0., 0.], dtype=float32)},
+                 'kernel': {'raw_value': Array([[ 0.21010089,  0.8289361 ,  0.04589564,  0.5422644 ],
+           [ 0.41914317,  0.84359694, -0.47937787, -0.49135214],
+           [-0.46072108,  0.4630125 ,  0.39276958, -0.9441406 ],
+           [-0.6690758 , -0.18474789, -0.57622856,  0.4821079 ]],      dtype=float32)}}}
+
+
+## Partial initialization
+
+In some cases (e.g., LoRA), you might want to randomly-initialize only *part of* your model parameters. 
+
+This can be done naively (initialize the whole model, then swap pre-trained params in), but that allocates additional memory. You can use `jax.jit`'s efficiently compiled code to make sure only params you need are initialized.
+
+
+```python
+# Some pretrained model state
+old_state = nnx.state(TwoLayerMLP(4, rngs=nnx.Rngs(0)))
+
+# Simple way to add a LoRA - not memory efficient
+simple_model = nnx.eval_shape(lambda: TwoLayerMLP(4, rngs=nnx.Rngs(42)))
+# On this line, extra kernel and bias is created inside the new LoRALinear!
+# They are wasted b/c we are to use the kernel and bias in `old_state` anyway.
+simple_model.linear1 = nnx.LoRALinear(4, 4, lora_rank=3, rngs=nnx.Rngs(42))
+nnx.update(simple_model, old_state)
+
+
+# Use jax.jit to automatically skip unused arrays - memory efficient!
+def get_lora_model():
+  model = TwoLayerMLP(4, rngs=nnx.Rngs(42))
+  model.linear1 = nnx.LoRALinear(4, 4, lora_rank=3, rngs=nnx.Rngs(42))
+  return model
+
+@functools.partial(jax.jit, donate_argnums=0, static_argnums=1)
+def partial_init(old_state, get_model):
+  new_state = nnx.state(get_model())
+  return nnx.State.merge(new_state, old_state)
+
+good_model = nnx.eval_shape(get_lora_model)
+# Note that `old_state` will be deleted after this `partial_init` call.
+new_state = partial_init(old_state, get_lora_model)
+nnx.update(good_model, new_state)
+
+# They are equivalent. The way RNGs are propagated in the new model is also preserved.
+np.testing.assert_allclose(simple_model(x), good_model(x))
```

Comment:
I'm a bit confused by this section, it seems to be doing the same operation in two different ways. I was thinking we could show case the use of `nnx.jit` here and use something like:

```python
old_state = nnx.state(TwoLayerMLP(4, rngs=nnx.Rngs(0)))

@functools.partial(nnx.jit, donate_argnums=0, static_argnums=1)
def partial_init(old_state, rngs):
  model = TwoLayerMLP(4, rngs=rngs)
  # add existing create
  nnx.update(model, old_state)
  # create new state
  model.linear1 = nnx.LoRALinear(4, 4, lora_rank=3, rngs=rngs))
  return model
  
  model = partial_init(old_state, nnx.Rngs(0))
```

Reply:
Hmm, for some reason the code you proposed didn't really save memory.

I added some lines to explain this section and also showed how to print the memory cost at each step. You can see that the `jax.jit` cases skipped the creation of extra unused params in `LoRALinear`. I don't know why your `nnx.jit` version doesn't repro this...

You can run this piece of code in a standalone python file to verify what I've seen. If `nnx.jit` can do this I'm happy to switch to `nnx.jit` instead.

Reply:
My bad, I put `nnx.update` in the middle but it should be at the end else the `model.linear1 =` assignment overrides the updated state. Here is the fixed code:

```python
  class TwoLayerMLP(nnx.Module):
      def __init__(self, dim, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(dim, dim, rngs=rngs)
        self.linear2 = nnx.Linear(dim, dim, rngs=rngs)

      def __call__(self, x):
        x = self.linear1(x)
        return self.linear2(x)

    old_state = nnx.state(TwoLayerMLP(4, rngs=nnx.Rngs(0)))

    print(f'Arrays start: {len(jax.live_arrays())}')
    print(len(jax.tree.leaves(old_state)))

    @partial(nnx.jit, donate_argnums=0, static_argnums=1)
    def partial_init(old_state, rngs):
      model = TwoLayerMLP(4, rngs=rngs)
      # create new state
      model.linear1 = nnx.LoRALinear(4, 4, lora_rank=3, rngs=rngs)
      # add existing create
      nnx.update(model, old_state)
      return model

    model = partial_init(old_state, nnx.Rngs(0))

    print(f'Arrays end: {len(jax.live_arrays())}')
    print(len(jax.tree.leaves(nnx.state(model))))
```

Reply:
Thanks! Added your code and replaced the `jax.jit` code. Thank you!